src/snowflake/snowpark/_internal/analyzer/metadata_utils.py

```diff
@@ -22,31 +56,61 @@ def infer_metadata(
     from snowflake.snowpark._internal.analyzer.snowflake_plan import SnowflakePlan
     from snowflake.snowpark._internal.analyzer.unary_plan_node import (
         Filter,
+        Project,
         Sample,
         Sort,
     )
 
     attributes = None
+    quoted_identifiers = None
     # If source_plan is a LogicalPlan, SQL simplifier is not enabled
     # so we can try to infer the metadata from its child (SnowflakePlan)
     # When source_plan is Filter, Sort, Limit, Sample, metadata won't be changed
     # so we can use the metadata from its child directly
     if isinstance(source_plan, (Filter, Sort, Limit, Sample)):
         if isinstance(source_plan.child, SnowflakePlan):
             attributes = source_plan.child._attributes
+            quoted_identifiers = source_plan.child._quoted_identifiers
+    elif isinstance(source_plan, Project):
+        quoted_identifiers = infer_quoted_identifiers_from_expressions(
+            source_plan.project_list,  # type: ignore
+            analyzer,
+            df_aliased_col_name_to_real_col_name,
+        )
     # If source_plan is a SelectStatement, SQL simplifier is enabled
     elif isinstance(source_plan, SelectStatement):
         # When attributes is cached on source_plan, just use it
         if source_plan._attributes is not None:
             attributes = source_plan._attributes
+        # When _column_states.projection is available, just use it, which is
+        # 1) cached on self._snowflake_plan._quoted_identifiers
```

Comment:
I don't understand the relationship between (1) and (2). Is it OR, or AND? If it's OR, then can we clarify in the comment? If it's AND, then how come it's both cached and inferred at the same time?

Reply:
It's OR, clarified in comments